Allow script plugins to be applied via the plugins DSL

eskatos opened the pull request:
```
plugins {
  script "path/to/other.gradle"
  script "https://example.com/another.gradle"
}
```

Local file paths are expressed as relative from the requesting script.
Remote URLs are also supported and cached.

Constrain plugins {} block api to distinguish binary and script plugin requests by both strongly modeled types and ad-hoc groovy parsing of the plugins block. This pave the way for proper support of the plugins {} block with the Gradle Kotlin DSL.

Implement ScriptPlugin PluginResolver by generating a synthetic "imperative" Plugin class that loads and applies the script plugin when applied.

This synthetic generated loader class and the script plugins are loaded into the buildSrc classloader scope.

Only Project targets are supported.
`apply false` is not supported.

See https://github.com/gradle/kotlin-dsl/issues/424

`Branch Build Accept` passed on CI: https://builds.gradle.org/viewLog.html?buildId=7908667

adammurdoch commented on subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/use/resolve/service/ScriptPluginUseIntegrationTest.groovy:
```diff
@@ -0,0 +1,412 @@
+
+            plugins {
+                script("gradle/hello.gradle")
+                script("gradle/../gradle/hello.gradle")
```
this should really fail given that we disallow duplicate plugin requests, as this two requests for the same thing.

bamboo replied:
It's like using two different ids to refer to the same plugin class. The requests are different but point at the same thing.

adammurdoch commented on subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/use/resolve/service/ScriptPluginUseIntegrationTest.groovy:
```diff
@@ -0,0 +1,412 @@
+        operations.all(ApplyPluginBuildOperationType) { it.displayName.contains("hello.gradle") }.size() == 1
+    }
+
+    def "same script plugin requested from different URLs gets applied once"() {
```
I'm not sure about using the content of the script as its identity. I would probably expect the location be used instead.

bamboo replied:
We will revisit this decision and take both designs into consideration again.

adammurdoch commented on subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/use/resolve/service/ScriptPluginUseIntegrationTest.groovy:
```diff
@@ -0,0 +1,412 @@
+    }
+
+
+    def "reasonable error message when the same script plugin request is done twice"() {
```
I don't really see the value in this case, given the other duplicate cases above that we ignore. I'd probably just remove this validation.

bamboo replied:
It's consistent with how Gradle behaves with plugin id requests. We should revisit the duplicate request handling strategy in a separate PR.

adammurdoch commented on subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/use/resolve/service/ScriptPluginUseIntegrationTest.groovy:
```diff
@@ -0,0 +1,412 @@
+import org.gradle.util.GradleVersion
+import org.junit.Rule
+
+class ScriptPluginUseIntegrationTest extends AbstractIntegrationSpec {
```
I think we should add some test coverage for class visibility:
- can the classes in the applying script see classes from the applied script? who else can see them? child projects? buildSrc classes? etc
- can the classes in the applied script see classes from the applying script? from buildSrc? from Gradle core? etc

adammurdoch replied:
Also, when a remote script is requested from multiple projects, that the script is downloaded once rather than several times.

adammurdoch replied:
There should be coverage for what happens when a plugin resolution strategy is being used and there are script plugin requests present:

- What does the resolution strategy see about the plugin request?
- What can the resolution strategy do and not do with these requests?
- What validation is applied?

bamboo replied:
👍

eskatos replied:
Classloading/visibility and caching of remote scripts across projects coverage added.
Resolution strategy related coverage still missing.

eskatos replied:
Coverage for plugin resolution strategies now added.

adammurdoch commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/use/internal/PluginResolverFactory.java:
```diff
@@ -67,26 +79,19 @@ public PluginResolver create() {
+     * Returns the default PluginResolvers used by Gradle. <p> The plugins will be searched in a chain from the first to the last until a plugin is found. So, order matters. <p> <ol> <li>{@link
+     * NoopPluginResolver} - Only used in tests.</li> <li>{@link CorePluginResolver} - distributed with Gradle</li> <li>{@link InjectedClasspathPluginResolver} - from a TestKit test's ClassPath</li>
+     * <li>Resolvers based on the entries of the `pluginRepositories` block</li> <li>{@link PluginResolutionServiceResolver} - from Gradle Plugin Portal if no `pluginRepositories` were defined</li>
+     * </ol> <p> This order is optimized for both performance and to allow resolvers earlier in the order to mask plugins which would have been found later in the order.
```
javadoc formatting is messed up.

bamboo replied:
Ouch.

adammurdoch commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/use/resolve/internal/ScriptPluginPluginResolver.java:
```diff
@@ -0,0 +1,272 @@
+        }
+    }
+
+    private static class ScriptPluginLoaderClassLoader extends URLClassLoader {
```
why do we need this?

bamboo replied:
Because we're generating the byte code for the classes representing the script plugins in memory. But your comment made us realize we should switch to generating it on disk and use `ClassLoaderScope#export` instead.

adammurdoch commented on subprojects/plugin-use/src/main/java/org/gradle/plugin/use/resolve/internal/ScriptPluginsScope.java:
```diff
@@ -0,0 +1,28 @@
+
+public class ScriptPluginsScope {
+
+    public static ClassLoaderScope from(ProjectRegistry<ProjectInternal> projectRegistry) {
```
We shouldn't use a static method here that makes assumptions about the arrangement of classloader scopes constructed elsewhere. Instead, should use a service that can explicitly answer the question about the class loader scopes, eg "give me the classloader scope for script plugins", "give me the classloader scope to use for buildSrc", etc.

This way the logic for the arrangement of the scopes is in one place (that service) and is explicit, rather than scattered around the place.

bamboo replied:
> We shouldn't use a static method here that makes assumptions about the arrangement of classloader scopes constructed elsewhere.

The function is just a way to avoid duplicating the logic temporarily as this will go away once script plugins get more sophisticated in subsequent stories.

adammurdoch commented on subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r40/PluginApplicationBuildProgressCrossVersionSpec.groovy:
```diff
@@ -24,7 +24,7 @@ import org.gradle.tooling.BuildException
 @ToolingApiVersion(">=2.5")
-@TargetGradleVersion(">=4.0")
+@TargetGradleVersion(">=4.0 <=4.1")
 class PluginApplicationBuildProgressCrossVersionSpec extends ToolingApiSpecification {
```
What changed so that these tests no longer apply? It looks like we duplicated a lot of test cases.

bamboo replied:
Build operation names for plugin applications changed.

bamboo commented on subprojects/core-api/src/main/java/org/gradle/plugin/management/PluginRequest.java:
```diff
@@ -35,6 +35,7 @@
     /**
      * The ID of the plugin requested. Never null.
      */
+    @Nullable
```
👍

bamboo commented on subprojects/core-api/src/main/java/org/gradle/plugin/management/PluginRequest.java:
```diff
@@ -44,6 +45,14 @@
+     * @since 4.2
+     */
+    @Nullable
+    String getScript();
```
At the request level this is only a string later resolved (in context) to a URI, just like plugin ids start as strings and later become `PluginId`.

This is also the DSL exposed to Kotlin (and to Groovy indirectly via AST transforms) and we definitely don't want people writing `script(uri("foo/bar.gradle.kts"))` or worse `script(new URI("foo/bar.gradle.kts"))`.

bamboo replied:
`URI` makes sense.

eskatos replied:
Done

eskatos replied:
`PluginRequest` is public API and is exposed to user code in plugin resolution strategies. Changing the type hierarchy may prove difficult or we'll have to live with the base type having all the properties that currently exists for binary plugins.

bamboo commented on subprojects/core-api/src/main/java/org/gradle/plugin/use/PluginDependenciesSpec.java:
```diff
@@ -117,4 +122,20 @@
+     * @return this
+     * @since 4.2
+     */
+    ScriptPluginDependencySpec script(String script);
```
See above.

eskatos replied:
Adding an overload accepting a `URI` would allow for Kotlin build scripts to do:
```kotlin
plugins {
  script(URI.create("some/file.gradle.kts"))
}
```

To support the same thing in Groovy we would have to enhance our ad-hoc interpreter for the `plugins` DSL to also support that overload. Otherwise Groovy users would be confused when reading the `PluginDependenciesSpec` DSL documentation.

eskatos replied:
Given `URI.create(..)` or anything creating a `URI` instance would be user code, the actual Groovy infrastructure makes it quite difficult (undesirable?) to support this.

My take is to **not** add a `URI` accepting overload to `PluginDependenciesSpec` for now.

/cc @adammurdoch, @bamboo

bamboo commented on subprojects/core-api/src/main/java/org/gradle/plugin/use/PluginDependencySpec.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2013 the original author or authors.
+ * Copyright 2010 the original author or authors.
```
Oops.

bamboo commented on subprojects/core/src/integTest/groovy/org/gradle/api/CrossBuildScriptCachingIntegrationSpec.groovy:
```diff
@@ -614,18 +614,6 @@ task fastTask { }
 
     def "script don't get recompiled if daemon disappears"() {
         root {
-            buildSrc {
```
It didn't seem to have anything to do with what's being tested. We reverted the change because it's unrelated to script plugins so it's confusing to be in this PR.

bamboo commented on subprojects/core/src/main/java/org/gradle/plugin/management/internal/DefaultPluginRequest.java:
```diff
@@ -25,49 +25,74 @@
 
 public class DefaultPluginRequest implements PluginRequestInternal {
```
Thanks. We'll look into that.

eskatos replied:
I'm not sure this would simplify serialization. We'll still want a single `PluginRequestSerializer` to be used in the Groovy `ScriptPluginFactory`. With split implementation it would have to rely on casting and still do null checks for some properties.

bamboo replied:
> With split implementation it would have to rely on casting 

This particular use of casting feels a lot like pattern matching against the different request variants so I would be ok with it from a design standpoint. From a performance standpoint I assume the implications of a downcast on such a shallow hierarchy are negligible. What would be your concern?

> and still do null checks for some properties

In the pattern matching / variant type interpretation, each branch would only have to check the nullability of the properties that make sense on the particular variant (`ScriptPluginRequest` or `BinaryPluginRequest`), right?

bamboo commented on subprojects/core/src/main/java/org/gradle/plugin/use/internal/PluginRequestCollector.java:
```diff
@@ -78,42 +82,62 @@ public PluginDependencySpec apply(boolean apply) {
+            Pair<String, Object> key = keyFor(request);
+            PluginRequestInternal existing = requestMap.get(key);
+            if (existing != null) {
+                InvalidPluginRequestException exception = new InvalidPluginRequestException(request, key.left + " '" + key.right + "' was already requested at line " + existing.getRequestingScriptLineNumber());
```
Let's review the duplicate request logic as a follow-up task.

bamboo commented on subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/use/PluginUseClassLoadingIntegrationSpec.groovy:
```diff
@@ -68,18 +68,15 @@ class PluginUseClassLoadingIntegrationSpec extends AbstractIntegrationSpec {
-                        getClass().getClassLoader().loadClass("org.gradle.test.TestPlugin")
-                        throw new AssertionError("plugin class *is* visible to build script")
-                    } catch (ClassNotFoundException expected) {}
+                    getClass().getClassLoader().loadClass("org.gradle.test.TestPlugin")
```
That's because the `PluginResolutionServiceTestServer` now returns `legacy = true` (see https://github.com/gradle/gradle/blob/381363a88958c209a0097d5a5393b8c0167133f9/subprojects/plugin-use/src/testFixtures/groovy/org/gradle/plugin/use/resolve/service/PluginResolutionServiceTestServer.groovy#L109) so we would get more coverage from tests on the class loading aspects we want to preserve.

bamboo commented on subprojects/plugin-use/src/integTest/groovy/org/gradle/plugin/use/PluginUseClassLoadingIntegrationSpec.groovy:
```diff
@@ -100,14 +97,9 @@ class PluginUseClassLoadingIntegrationSpec extends AbstractIntegrationSpec {
-            } catch (ClassNotFoundException ignore) {
-
-            }
+            getClass().getClassLoader().loadClass('org.gradle.api.plugins.JavaPlugin')
```
Same as above.

eskatos commented on subprojects/core/src/main/java/org/gradle/plugin/use/internal/PluginRequestCollector.java:
```diff
@@ -49,18 +50,21 @@ public PluginRequestCollector(ScriptSource scriptSource) {
     }
 
-    private static class DependencySpecImpl implements PluginDependencySpec {
+    private static class DependencySpecImpl implements PluginDependencySpec, ScriptPluginDependencySpec {
```
Implementation for `ScriptPluginDependencySpec` would be empty so I'd say it's not worth it at this stage. We can split the implementations when we introduce `apply(false)` support for script plugins.

eskatos commented:
Here is a summary of unaddressed review comments.

### Release notes and documentation

See [comments](https://github.com/gradle/gradle/pull/2825#discussion_r136212366).
I'll address this once the other things are done / agreed on.

### `PluginDependenciesSpec.script` overload taking a `URI`

See [comments](https://github.com/gradle/gradle/pull/2825#discussion_r136212731).
I say we should not do this just yet.

### Rework the `PluginDependencySpec` type hierarchy

This is mostly about getting meaningful names.
We have a plan for this and it will be implemented.

### Split the `PluginDependencySpec` implementation

See [comments](https://github.com/gradle/gradle/pull/2825#discussion_r136215266).
I say we should not do this just yet.

### Split the `PluginRequest` type hierarchy

See [comments](https://github.com/gradle/gradle/pull/2825#discussion_r136212305).
My main concern here is backwards compatibility. My take would be to keep what we have in this PR.

### Split the `PluginRequest` implementation

See [comments](https://github.com/gradle/gradle/pull/2825#discussion_r136214556).
I say we should not do this now.

@adammurdoch, @bamboo, could you comment on this, or simply ACK?

bamboo commented:
Hey @eskatos, thanks for the summary.

I slightly disagree on `Split the PluginRequest implementation` and posted a specific comment. I agree with all the other points.

eskatos commented:
I ended up splitting both `PluginDependencySpec` and `PluginRequest` implementations.

Next step, rework the `PluginDependencySpec` public type hierarchy.

eskatos commented:
[Branch Build Accept](https://builds.gradle.org/viewLog.html?buildId=8324800&) passed before the last two documentation only commits.

I just triggered another [Branch Build Accept](https://builds.gradle.org/viewQueued.html?itemId=8329455) with these last changes.

eskatos commented:
@bamboo, @adammurdoch could you have another look?

oehme commented:
@eskatos do we still plan to have this in 4.3?

eskatos commented:
@oehme no, it's on hold, I'll update the milestone right away, thanks for the heads up!
